Replay st elements inside memo/singleton functions

AnOctopus opened the pull request:
Still in progress, pending completion of block handling

## 📚 Context

- What kind of change does this PR introduce?

  - [ ] Bugfix
  - [x] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧠 Description of Changes

- Record elements produced during execution of a memo/singleton decorated function, and replay them when we skip execution because the value was cached
- Stores messages for memo as part of the pickled value blob, and uses a separate dict to store singleton's messages
- Removed caching warning for element functions that now work with caching decorators

  - [ ] This is a breaking API change
  - [ ] This is a visible (user-facing) change


## 🧪 Testing Done

- [ ] Screenshots included
- [x] Added/Updated unit tests
- [ ] Added/Updated e2e tests

AnOctopus committed:
Let memo not warn about non-widget elements

AnOctopus committed:
Let singleton also not warn about nonwidgets

AnOctopus committed:
In-memory saving and replay of elements for memo/singleton

AnOctopus committed:
On-disk saving/retrieval of memo messages

AnOctopus commented on lib/streamlit/elements/__init__.py:
```diff
@@ -11,3 +11,53 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
```
Need to go over these lists to make sure they're accurate and comprehensive

kajarenc replied:
noticed that `st.video` is not present in either list

AnOctopus committed:
Fixup call stack class documentation

AnOctopus committed:
Call assertion in test

AnOctopus commented on lib/streamlit/caching/__init__.py:
```diff
@@ -14,8 +14,25 @@
 
-from .memo_decorator import MEMO_CALL_STACK, _memo_caches, MemoAPI
-from .singleton_decorator import SINGLETON_CALL_STACK, _singleton_caches, SingletonAPI
+from google.protobuf.message import Message
```
I noticed that DeltaGenerator only imports this during typechecking time, and probably I should do the same with all these new instances of importing it

AnOctopus replied:
Upon looking more, we're importing it via all our protobuf imports anyway, so it should be fine to leave it as a regular import

AnOctopus commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -38,11 +42,17 @@
 _LOGGER = get_logger(__name__)
 
 
+@define
```
I'd like to get opinions about using new-style attrs vs old-style attrs vs dataclasses

AnOctopus commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -38,11 +42,17 @@
 
     @abstractmethod
-    def read_value(self, value_key: str) -> Any:
+    def read_value(self, value_key: str) -> Tuple[Any, List[MsgData]]:
```
CachedValue could be moved into this module and this method changed to return it

AnOctopus commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -123,20 +137,26 @@ def get_or_create_cached_value():
-                return_value = cache.read_value(value_key)
+                return_value, messages = cache.read_value(value_key)
                 _LOGGER.debug("Cache hit: %s", func)
+                dg = st._main
```
Will this always be up to date wrt blocks, delta paths?

AnOctopus replied:
I think it _almost_ works, except the sidebar is a distinct DG rather than a different state the main DG happens to be in, so it would render to the wrong place if you called a function in the sidebar in a cached fn. The st functions themselves interact with the correct DG because they're invoked on the DG, but this code doesn't have that luxury. Instead, we need to save the reference to the dg when we save the message, like how the warning function uses the active dg, except we need to keep it around instead of using it right then.

I went back and forth on if that would work, and I think it does, because whichever dg is used to invoke the element fn the first time, it should be as if we used that to invoke it again when reading from the cache.

Oh no, do we need to keep separate message stacks for sidebar vs main? I think we might

AnOctopus commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -228,6 +252,40 @@ def _show_cached_st_function_warning(
             dg.exception(e)
 
 
+class CachedFunctionMessagesCallStack(threading.local):
```
This name isn't great, especially with CachedFunctionCallStack still hanging around. One or both should get a more informative name

AnOctopus commented on lib/streamlit/caching/memo_decorator.py:
```diff
@@ -456,15 +477,15 @@ def _read_from_disk_cache(self, key: str) -> bytes:
             raise CacheError("Unable to read from cache") from e
 
-    def _write_to_mem_cache(self, key: str, pickled_value: bytes) -> None:
+    def _write_to_mem_cache(self, key: str, value: bytes) -> None:
```
This is a pretty pointless rename, the aftermath of a series of refactors as I tried to figure out what types these things should interact via

AnOctopus commented on lib/streamlit/caching/singleton_decorator.py:
```diff
@@ -248,23 +256,26 @@ def __init__(self, key: str, display_name: str):
         self.display_name = display_name
         self._mem_cache: Dict[str, Any] = {}
         self._mem_cache_lock = threading.Lock()
+        self._message_cache: Dict[str, List[MsgData]] = {}
```
I originally had both memo and singleton with separate dicts for values and messages, but changed memo to make persisting to disk easier. Potentially I should propagate that change to singleton as well, to reduce unnecessary differences between them

AnOctopus committed:
Move CachedValue to cache_utils, use as return type for read_value

Unused import cleanup

AnOctopus commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -38,11 +42,23 @@
+
+
+@define
+class CachedValue:
```
CachedValue is an awkward name, since the point is that it's the value _plus_ some message stuff

AnOctopus replied:
Oh, maybe `CachedResult`

AnOctopus commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -38,11 +42,23 @@
     @abstractmethod
-    def read_value(self, value_key: str) -> Any:
+    def read_value(self, value_key: str) -> CachedValue:
         """Read a value from the cache.
```
Update to also mention messages, for here and related following methods

AnOctopus commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -123,20 +143,27 @@ def get_or_create_cached_value():
 
             try:
-                return_value = cache.read_value(value_key)
+                value = cache.read_value(value_key)
```
value is a dumb name, both because it has messages and because it leads to the following `value.value`

AnOctopus commented on lib/streamlit/caching/memo_decorator.py:
```diff
@@ -435,9 +449,8 @@ def clear(self) -> None:
     def _read_from_mem_cache(self, key: str) -> bytes:
```
Should add some documentation explaining why these work with bytes rather than a structured type (avoid unneeded roundtripping when saving to mem a value from disk)

AnOctopus committed:
Import cleanup

AnOctopus committed:
CachedValue -> CachedResult

AnOctopus committed:
Fix mypy complaints

AnOctopus committed:
Revert meaningless parameter name changes

AnOctopus commented:
Making ready for review in case someone wants to jump in and review it while I'm fixing stuff I've already pointed out

AnOctopus committed:
Replace attrs use with dataclass

Following discussion with other maintainers

AnOctopus committed:
Mention messages in Cache method docstrings

AnOctopus committed:
Add comment about returning/taking bytes from some fns

AnOctopus committed:
Rename cache call stacks for specificity

AnOctopus committed:
Rename Cache call stack property to match new name scheme

AnOctopus commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -38,12 +42,24 @@
     @abstractmethod
-    def read_value(self, value_key: str) -> Any:
-        """Read a value from the cache.
+    def read_value(self, value_key: str) -> CachedResult:
```
Rename these to read/write _result, to match new convention

AnOctopus commented on lib/streamlit/caching/memo_decorator.py:
```diff
@@ -392,36 +401,41 @@ def get_stats(self) -> List[CacheStat]:
         try:
-            return pickle.loads(pickled_value)
+            entry = pickle.loads(pickled_entry)
+            assert isinstance(entry, CachedResult)
```
This should just be a nonfatal check that throws a CacheError as if there was an UnpicklingError

AnOctopus committed:
Rename Cache methods to "result" convention

AnOctopus committed:
Throw CacheError if result unpickled as something unexpected

AnOctopus committed:
Use combined result cache for singleton

AnOctopus committed:
Save WIP so I can show Emily

Hi emily I love you

AnOctopus committed:
Basic test cases passing

AnOctopus committed:
Remove spurious prints

AnOctopus committed:
Save main and sidebar ids in CachedResults

AnOctopus committed:
Clean up comments

AnOctopus committed:
Add note on rewriting DG invocations

AnOctopus commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -35,15 +51,44 @@
 _LOGGER = get_logger(__name__)
 
 
+@dataclass
```
Add docstrings explaining these

AnOctopus commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -85,14 +130,51 @@ def cache_type(self) -> CacheType:
         raise NotImplementedError
 
 
+def replay_result_messages(result: CachedResult) -> None:
```
Raise an explanatory error instead of letting the KeyNotFound bubble up

willhuang1997 replied:
When does this get called? it gets called when the key to a widget is not found?

AnOctopus replied:
It gets called when a function's result was found in the cache

AnOctopus commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -228,6 +321,97 @@ def _show_cached_st_function_warning(
+"""
+
+
+class CacheMessagesCallStack(threading.local):
```
Consider moving most of this work into the DG class itself

willhuang1997 replied:
Is there a reason why to move it into the DG Class? If so, can you put them in the comments? or what would be the pros and cons of each approach?

AnOctopus replied:
The logic of replaying elements is very similar to the process of generating them in the first place, and having that knowledge in one place is better than having it spread between two (one of which requires understanding the other anyway), all else equal.

AnOctopus commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -228,6 +321,97 @@ def _show_cached_st_function_warning(
+        for s in self._seen_dg_stack:
+            s.add(returned_dg_id)
+
+    def get_dg_to_save_as_invoked(self, invoked_id: str, acting_on_id: str) -> str:
```
This needs a better name, especially since one of the parameters is described as "invoked"

AnOctopus commented on lib/streamlit/delta_generator.py:
```diff
@@ -523,6 +523,15 @@ def _enqueue(
+        caching.save_element_message(
+            delta_type,
+            element_proto,
+            invoked_dg_id=str(id(self)),
+            used_dg_id=str(id(dg)),
+            returned_dg_id=str(id(output_dg)),
```
The id should probably be a property so we don't have to write str(id(...)) everywhere

AnOctopus commented on lib/tests/streamlit/caching/common_cache_test.py:
```diff
@@ -255,6 +264,225 @@ def cached_widget():
+            ("singleton", singleton),
+        ]
+    )
+    def test_cached_st_function_replay_sidebar(self, _, cache_decorator):
```
It would be really nice to have some helper function to turn deltas into a tree to compare against in tests

AnOctopus commented on lib/tests/streamlit/caching/common_cache_test.py:
```diff
@@ -255,6 +264,225 @@ def cached_widget():
+        foo(1)
+
+        deltas = self.get_all_deltas_from_queue()
+        text = [
+            element.text.body
+            for element in (delta.new_element for delta in deltas)
+            if element.WhichOneof("type") == "text"
+        ]
```
Extract this

AnOctopus commented on lib/tests/streamlit/caching/common_cache_test.py:
```diff
@@ -255,6 +264,225 @@ def cached_widget():
+        ]
+        assert text == ["1", "---", "1"]
+
+        paths = [
+            msg.metadata.delta_path
+            for msg in self.forward_msg_queue._queue
+            if msg.HasField("delta")
+        ]
```
Extract this

AnOctopus committed:
Add id property for DG

AnOctopus committed:
Improve name and docs of DG id selector function

AnOctopus committed:
Clean up cache test helpers

AnOctopus committed:
Add some description of cache dataclasses

willhuang1997 commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -35,15 +51,52 @@
 
+@dataclass
+class ElementMsgData:
+    """A non-interactive element's message and related metadata for
```
This is specifically for elements that are non-interactive or are you saying this msgData is not interactive?

AnOctopus replied:
non-interactive elements. I wish we had a term specifically for them

kajarenc replied:
> non-interactive elements. I wish we had a term specifically for them

Maybe "static elements", or is it too general word for it?

Another suggestion is "smart elements" / "dumb elements" in analogy with "smart components" / "dumb components" in React

AnOctopus replied:
"static elements" kinda works, but feels at odds with some of them being layout blocks that you actively change in the script

willhuang1997 commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -85,14 +138,51 @@ def cache_type(self) -> CacheType:
+    by an st function call. Then, for each recorded message, we get the current DG
+    instance corresponding to the DG the message was originally called on, and enqueue
+    the message using that, recording any new DGs produced in case a later st function
+    call is on one of them.
```
So it looks like this comment is saying that we record BOTH the identifier for the dg that was called on (does that mean st.foo?) and then the dg returned by an st function call (when does this happen?) and then we enqueue each of those messages again and in each of those messages, we keep track of any new dgs? I'm personally confused about this method but maybe because im lacking on knowledge on dg and cache.

AnOctopus replied:
`st.foo` is implicitly a call on the main DG, but st functions can also be called on other DGs like the sidebar, a container, etc. The main reason a dg is returned from an st function is for layout blocks, but anything that doesn't have a real return value (like st.write, st.empty) will actually return a dg so you can do weird layout stuff with them

willhuang1997 commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -228,6 +329,106 @@ def _show_cached_st_function_warning(
+
+Implicit invocation on a known DG -> Explicit invocation on that DG
+Implicit invocation on an unknown DG -> Rewrite as explicit invocation on main
+    This is situation 2 above, and the DG is a block entirely outside our function call,
```
So far I understand that this is situation 2 but I'm confused what is main and how that interacts here.

AnOctopus replied:
main is the implicit global DG. when you call `st.foo`, you are actually calling a method on main

willhuang1997 commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -228,6 +329,106 @@ def _show_cached_st_function_warning(
+Implicit invocation on a known DG -> Explicit invocation on that DG
+Implicit invocation on an unknown DG -> Rewrite as explicit invocation on main
+    This is situation 2 above, and the DG is a block entirely outside our function call,
+    so we interpret it as "put these elements within the current context", which is achieved
```
Also, "put these elements within the current context" doesn't really make any sense to me. Does that mean we just need to make sure that we put some element within a st.block? St.blocks are st.containers and vertical blocks?

AnOctopus replied:
When you call st.foo, it is technically a call on the main DG. But if you're in a `with st.some_block_thing` block, main will look up the DG for that and use it instead of itself for handling the call. So `st.foo` actually means "call this on the dg of the current block context", (treating main as an all-encompassing block context, because it kinda is)

willhuang1997 commented on lib/streamlit/caching/cache_utils.py:
```diff
@@ -228,6 +329,106 @@ def _show_cached_st_function_warning(
+
+        See Note [DeltaGenerator method invocation]
+
+        invoked_id is the DG the st function was called on, usually `st._main`
```
would be nice to have a period at end here.

kajarenc commented:
@AnOctopus look's good to me! 

could you please give me access to your repo
I want to manually test something

I have some fears (maybe unfounded!) that with widgets that interact with the file system (`st.image` , `st.download_button` ) there may be problems, since if I remember correctly we put data to InMemoryFileManager on each rerun and put only URL to proto, and if the result of the execution is cached, the link may be broken.

AnOctopus commented:
@kajarenc You should have read access to my repo already, since it should be public. Can you not access it, or do you need write access for some reason?

Widgets in cached functions are still unsupported. It sounds like you're concerned about widgets having their results cached resulting in incorrect values, is that right?

kajarenc commented:
oops, @AnOctopus , yep!
Just forgot I don't need push access to just pull your branch :) :) Read access more than enough

I mostly wanted to test `st.image` since it is in `NONWIDGET_ELEMENTS` list, but using it inside cached function shows warning (Btw with strange line `Move the *st.imgs()* or st.write() call outside foo()` 


```python
import streamlit as st
from PIL import Image


@st.experimental_memo
def foo():
    image = Image.open('dog.jpeg')
    st.image(image, caption='DOG DOG')

foo()

x = st.slider("ADD INTERACTIVITY", 0, 100, 27)
st.write(x ** 2)

```

AnOctopus committed:
Improve doc explanation of DG invocation rewrites

AnOctopus committed:
Create Error for invalid element replay

It would be better to be able to specify the st function that triggered
it, but we aren't currently tracking that in the callstack and I don't
want to retrofit it to do that right now.

AnOctopus commented:
Possibly the issue is that `st.image` should be treated as a widget and I just put it in the wrong list the first time

AnOctopus committed:
Include st.image in cache warning

AnOctopus commented:
@kajarenc Do you know if the filesystem link issue applies to other things, like media (audio/video)?

kajarenc commented:
@AnOctopus  That filesystem link issue isn't proved by me, it was just my assumption (since another error was raised during my testing)
But if it exists, I think it should work the same way with `st.audio`, `st.video` and `st.download_button`

AnOctopus committed:
Exclude media elements from replay

AnOctopus committed:
Use specific Error type in cache test

AnOctopus committed:
Call function in test

AnOctopus committed:
Update lists of elements
